Added `FunctorChangeFunctorMaterial`

joshuahansel opened the pull request:
- Added `FunctorChangeFunctorMaterial`, which computes the change in a functor over a time step, nonlinear iteration, or fixed point iteration.
- Added `FixedPoint` as a possible value in `IterationType`
- Added a common vector tag for the previous fixed point iteration solution
- Used this in fixed point solves and refactored
- Fixed miscellaneous small things
- Added `parallel_type` argument to `needSolutionState` (default to `GHOSTED`)
- Added `solutionStateParallelType()` for checking parallel type in `needSolutionState` in case vector was requested with different type already. An error is thrown if already declared with `PARALLEL` but requested to upgrade to `GHOSTED`, but not the opposite, which is fine.
- Moved the call of `allocateStorage(true)` from fixed point solve constructors to `initialSetup()` to allow `PARALLEL` state to be requested after any objects request `GHOSTED` for the previous fixed point solution vector.

Closes #31030

joshuahansel committed:
Made FEProblemBase::needSolutionState consistent

FEProblemBase::needSolutionState(n) was actually only doing
SystemBase::needSolutionState(n-1). Also you don't need to
call SystemBase::needSolutionState for every state, just the max,
since the lower states get added internally.

Refs #31030

joshuahansel committed:
Fixed Steffensen doc typo

joshuahansel committed:
Removed unused data members in some PPs

joshuahansel committed:
Added FixedPoint IterationType and previous FP vector tag

Refs #31030

joshuahansel committed:
Refactored fixed point solves to use common vector tag

Refs #31030

joshuahansel committed:
Added FunctorChangeFunctorMaterial

Closes #31030

joshuahansel committed:
Added operator== for StateArg

Refs #31030

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3106153/), step Docs: sync website on f4af8f5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31233/site)

This comment will be updated on new commits.

joshuahansel commented on framework/src/functormaterials/FunctorChangeFunctorMaterial.C:
```diff
@@ -0,0 +1,89 @@
+Moose::StateArg
+FunctorChangeFunctorMaterialTempl<is_ad>::referenceState(const MooseEnum & change_over) const
+{
+  if (change_over == "time_step")
```
This is only called once, in the constructor.

lindsayad replied:
yea I think this is fine for a one-time comparison

joshuahansel committed:
Apply suggestions from code review

- Added link to fixed point iterations
- Added "the" in parameter description
- Removed method description from .C file
- Removed unnecessary curly braces

Refs #31030

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

GiudGiud commented on framework/src/systems/SystemBase.C:
```diff
@@ -1451,7 +1470,9 @@ SystemBase::needSolutionState(const unsigned int state,
       auto tag = _subproblem.addVectorTag(oldSolutionStateVectorName(i, iteration_type),
                                           Moose::VECTOR_TAG_SOLUTION);
-      solution_states[i] = &addVector(tag, true, GHOSTED);
+      const ParallelType parallel_type =
+          iteration_type == Moose::SolutionIterationType::FixedPoint ? PARALLEL : GHOSTED;
+      solution_states[i] = &addVector(tag, true, parallel_type);
```
whatever I was doing in the FixedPointSolve, copying solution vectors, was it keeping them ghosted? or was it a ghosted to parallel copy

joshuahansel replied:
That's a good point. I believe there's a GHOSTED-to-PARALLEL copy in those already - I don't see what's different there

lindsayad replied:
Move parallel to ghosted and ghosted to parallel copying are supported

joshuahansel committed:
Added parallel type to needSolutionState

Refs #31030

joshuahansel commented:
This should be ready. See last 3 bullets for how I addressed the PARALLEL vs. GHOSTED issue.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3125104/), step Generate coverage on 3658d29 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>e06439</th>
      <th colspan=4>#31233 3658d2</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.57%</td>
      <td>85.57%</td>
      <td>+0.00%</td>
      <td>94.87%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>116379</td>
      <td>116426</td>
      <td>+47</td>
      <td>111</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19631</td>
      <td>19635</td>
      <td>+4</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31233/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31233/coverage/framework)


## Modules coverage

Inconsistent report tags were found between the head and base reports.
This can happen when reports are missing from either the head or the base.

Inconsistent tags:
  module-chemical_reactions
  module-chemical_reactions-parallel
  module-chemical_reactions-recover
  module-chemical_reactions-restep
  module-combined
  module-combined-recover
  module-combined-restep
  module-contact
  module-contact-parallel
  module-contact-recover
  module-contact-restep
  module-electromagnetics
  module-electromagnetics-parallel
  module-electromagnetics-recover
  module-electromagnetics-restep
  module-external_petsc_solver
  module-external_petsc_solver-parallel
  module-external_petsc_solver-recover
  module-external_petsc_solver-restep
  module-fluid_properties
  module-fluid_properties-parallel
  module-fluid_properties-recover
  module-fluid_properties-restep
  module-fsi
  module-fsi-parallel
  module-fsi-recover
  module-fsi-restep
  module-functional_expansion_tools
  module-functional_expansion_tools-parallel
  module-functional_expansion_tools-recover
  module-functional_expansion_tools-restep
  module-geochemistry
  module-geochemistry-parallel
  module-geochemistry-restep
  module-heat_transfer
  module-heat_transfer-parallel
  module-heat_transfer-recover
  module-heat_transfer-restep
  module-level_set
  module-level_set-parallel
  module-level_set-restep
  module-misc
  module-misc-parallel
  module-misc-recover
  module-misc-restep
  module-navier_stokes
  module-navier_stokes-parallel
  module-navier_stokes-recover
  module-navier_stokes-restep
  module-optimization
  module-optimization-parallel
  module-optimization-recover
  module-optimization-restep
  module-peridynamics
  module-peridynamics-parallel
  module-peridynamics-restep
  module-phase_field
  module-phase_field-parallel
  module-phase_field-recover
  module-phase_field-restep
  module-porous_flow
  module-porous_flow-parallel
  module-porous_flow-recover
  module-porous_flow-restep
  module-ray_tracing
  module-ray_tracing-parallel
  module-ray_tracing-recover
  module-ray_tracing-restep
  module-rdg
  module-rdg-parallel
  module-rdg-restep
  module-reactor
  module-reactor-parallel
  module-reactor-recover
  module-reactor-restep
  module-richards
  module-richards-restep
  module-scalar_transport
  module-scalar_transport-parallel
  module-scalar_transport-recover
  module-scalar_transport-restep
  module-solid_mechanics
  module-solid_mechanics-parallel
  module-solid_mechanics-restep
  module-solid_properties
  module-solid_properties-parallel
  module-solid_properties-recover
  module-solid_properties-restep
  module-stochastic_tools
  module-stochastic_tools-parallel
  module-stochastic_tools-restep
  module-subchannel
  module-subchannel-parallel
  module-subchannel-recover
  module-subchannel-restep
  module-thermal_hydraulics
  module-thermal_hydraulics-parallel
  module-thermal_hydraulics-recover
  module-thermal_hydraulics-restep
  module-xfem
  module-xfem-parallel
  module-xfem-restep

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31233/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad reviewed (commented):
7/27 through but done for the day

joshuahansel commented on framework/include/executioners/FixedPointSolve.h:
```diff
@@ -121,6 +121,14 @@ class FixedPointSolve : public SolveObject
+  /**
+   * Returns true if there is relaxation.
+   *
+   * @param primary Whether this routine is to save the variables for the primary transformed
+   *                quantities (as main app) or the secondary ones (as a subapp)
```
Sorry, that was copy and paste from the `saveVariables(primary)` doxygen. I'll change to "True if this is the parent app; false if this is a child app"

lindsayad commented on framework/include/systems/SystemBase.h:
```diff
@@ -224,12 +225,20 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
+   * Returns the parallel type of the given solution state
+   */
+  libMesh::ParallelType
+  solutionStateParallelType(const unsigned int state,
+                            const Moose::SolutionIterationType iteration_type) const;
```
Maybe this should just take a `StateArg`?

joshuahansel replied:
I suggest it be consistent with these:

- `needSolutionState`
- `solutionState`
- `oldSolutionStateVectorName`

So I can make them all `StateArg` or leave them be.

lindsayad commented on framework/include/systems/SystemBase.h:
```diff
@@ -1072,7 +1081,7 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
 
   /// The solution states (0 = current, 1 = old, 2 = older, etc)
-  std::array<std::vector<NumericVector<Number> *>, 2> _solution_states;
+  std::array<std::vector<NumericVector<Number> *>, 3> _solution_states;
```
this needs more doxygen with respect to what the index into the `std::array` is

lindsayad commented on framework/src/executioners/FixedPointSolve.C:
```diff
@@ -552,3 +559,12 @@ FixedPointSolve::autoAdvance() const
+FixedPointSolve::performingRelaxation(const bool primary) const
+{
+  if (primary)
+    return _relax_factor != 1.;
+  else
+    return _secondary_relaxation_factor != 1.;
```
I always pause at floating point comparisons. This is probably fine. I'll leave it to you whether to keep as-is or do a fuzzy equal comparison

lindsayad commented on framework/src/executioners/PicardSolve.C:
```diff
@@ -22,107 +22,88 @@ PicardSolve::validParams()
-  }
+  // Allocate storage for the previous postprocessor values
+  (*transformed_pps_values).resize((*transformed_pps).size());
+  for (size_t i = 0; i < (*transformed_pps).size(); i++)
```
```suggestion
  for (const auto i : index_range(*transformed_pps))
```

lindsayad commented on framework/src/executioners/PicardSolve.C:
```diff
@@ -22,107 +22,88 @@ PicardSolve::validParams()
-    for (size_t i = 0; i < (*transformed_pps).size(); i++)
-      (*transformed_pps_values)[i][0] = getPostprocessorValueByName((*transformed_pps)[i]);
+  // Save postprocessor previous values
+  for (size_t i = 0; i < (*transformed_pps).size(); i++)
```
```suggestion
  for (const auto i : index_range(*transformed_pps))
```

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -692,7 +692,7 @@ FEProblemBase::createTagSolutions()
     // the solution states containers so populating the zeroth state does not cost us the memory of
     // a new vector
-    needSolutionState(2, Moose::SolutionIterationType::Nonlinear);
+    needSolutionState(1, Moose::SolutionIterationType::Nonlinear);
```
lol what the hell. Nice fix. I wouldn't be surprised if git blame points to me on this one

lindsayad commented on test/tests/functormaterials/functor_change/tests:
```diff
@@ -0,0 +1,26 @@
+      type = CSVDiff
+      input = 'fp_parent.i'
+      csvdiff = 'fp_parent_out.csv'
+      detail = 'a fixed point iteration.'
```
```suggestion
      detail = 'a fixed point iteration,'
```

lindsayad commented on test/tests/functormaterials/functor_change/tests:
```diff
@@ -0,0 +1,26 @@
+      type = CSVDiff
+      input = 'time_step.i'
+      csvdiff = 'time_step_out.csv'
+      detail = 'a time step.'
```
```suggestion
      detail = 'a time step, or'
```

joshuahansel committed:
Apply suggestions from code review

- Fixed SQA test description detail
- Used better for-loop

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

joshuahansel committed:
Improved _solution_states doxygen

Refs #31030

joshuahansel committed:
Used fuzzy float comparison in FixedPointSolve

Refs #31030
